lib/models/database/post_repository.dart

```diff
@@ -0,0 +1,278 @@
+    try {
+      return PostLocationFirestore(
+        geoPoint: data[geoPointField],
+        geohash: data[geohashField],
+      );
+    } catch (e) {
+      if (e is TypeError) {
+        throw Exception("Cannot parse post location document: ${e.toString()}");
+      } else {
+        rethrow;
+      }
+    }
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other)) return true;
+
+    return other is PostLocationFirestore &&
+        other.geoPoint == geoPoint &&
+        other.geohash == geohash;
+  }
+
+  @override
+  int get hashCode => Object.hash(geoPoint, geohash);
+}
+
+class PostRepository {
+  final GeoFlutterFire _geoFire;
+  final GeoLocationService _geoLocationService;
+
+  static const collectionName = "posts";
+  final CollectionReference _collectionRef;
+
+  static const double kmPostRadius = 0.1;
+
+  PostRepository({
+    required FirebaseFirestore firestore,
+    required GeoFlutterFire geoFire,
+    required GeoLocationService geoLocationService,
+  })  : _collectionRef = firestore.collection(collectionName),
+        _geoFire = geoFire,
+        _geoLocationService = geoLocationService;
+
+  /// Get the posts near the current location of the user
+  /// The radius is defined by [kmPostRadius]
+  Future<List<PostFirestore>> getNearPosts() async {
+    final point = await _getCurrentGeoFirePoint();
+
+    return _getPostsNearLocation(point, kmPostRadius);
+  }
+
+  /// Get a post by its id
+  Future<PostFirestore> getPost(PostFirestoreId postId) async {
+    final docSnap = await _collectionRef.doc(postId).get();
+
+    return PostFirestore.fromDb(docSnap);
+  }
+
+  /// Adds a post at the current location of the user
+  Future<void> addPost(PostFirestoreData postData) async {
+    final point = await _getCurrentGeoFirePoint();
+
+    return _addPostAtLocation(point, postData);
+  }
+
+  /// Deletes a post by its id
+  Future<void> deletePost(PostFirestoreId postId) async {
+    await _collectionRef.doc(postId).delete();
+  }
+
+  Future<void> _addPostAtLocation(
+    GeoFirePoint point,
+    PostFirestoreData postData,
+  ) async {
+    // The `point.data` returns a Map<String, dynamic> consistent with the
+    // class [PostLocationFirestore]. This is because the field name values
+    // are hardcoded in the [GeoFlutterFire] library
+
+    await _collectionRef.add({
+      PostFirestore.locationField: point.data,
+      ...postData.toDbData(),
+    });
+  }
+
+  Future<List<PostFirestore>> _getPostsNearLocation(
+    GeoFirePoint point,
+    double radius,
+  ) async {
+    final posts = await _geoFire
+        .collection(collectionRef: _collectionRef)
+        .within(
+          center: point,
+          radius: radius,
+          field: PostFirestore.locationField,
+          strictMode: true,
+        )
+        .first;
+
+    return posts.map((docSnap) => PostFirestore.fromDb(docSnap)).where((post) {
+      // We need to filter the posts because the query is not exact
```

Comment:
Once again, because of the poor documentation, I digged through the code to find out that with strict mode, the `within` function would still add a 2% margin. So, I decided to add real strict filter in out function.

Code from the library:
```dart
      final filteredList = strictMode
          ? mappedList
              .where((DistanceDocSnapshot doc) =>
                      doc.distance! <=
                      // Here they add a 2% margin
                      radius * 1.02 // buffer for edge distances;
                  )
              .toList()
          : mappedList.toList();
```